### PR TITLE
Preserve join aliases in dashcard queries (fixes 25278 & 32557)

### LINKED ISSUE #25278: Automatic explorations `compare to the rest` option fails for questions with joined table data present in the result
**Describe the bug**
When a question is created having joined tables columns in the result the `compare to the rest` option fails for some cards with:

` Error processing query: Cannot determine the source table or query for Field clause [:field 30 {:join-alias ...`

**Logs**
<details><summary>Full stacktrace</summary>

```
metabase-postgres         | 2022-09-07 07:38:30,004 ERROR middleware.catch-exceptions :: Error processing query: Cannot determine the source table or query for Field clause [:field 30 {:join-alias "Products"}]
metabase-postgres         | {:database_id 1,
metabase-postgres         |  :started_at #t "2022-09-07T07:38:25.091233Z[GMT]",
metabase-postgres         |  :error_type :invalid-query,
metabase-postgres         |  :json_query
metabase-postgres         |  {:type "query",
metabase-postgres         |   :database 1,
metabase-postgres         |   :query
metabase-postgres         |   {:source-table 4,
metabase-postgres         |    :breakout [["field" 31 {:binning {:strategy "default"}}]],
metabase-postgres         |    :aggregation [["count"]],
metabase-postgres         |    :filter
metabase-postgres         |    ["and"
metabase-postgres         |     ["=" ["field" 36 nil] "680"]
metabase-postgres         |     ["=" ["field" 33 nil] "118"]
metabase-postgres         |     ["=" ["field" 31 nil] 3]
metabase-postgres         |     ["=" ["field" 30 {:join-alias "Products"}] "118"]]},
metabase-postgres         |   :parameters [],
metabase-postgres         |   :middleware {:js-int-to-string? true, :add-default-userland-constraints? true}},
metabase-postgres         |  :native nil,
metabase-postgres         |  :status :failed,
metabase-postgres         |  :class clojure.lang.ExceptionInfo,
metabase-postgres         |  :stacktrace
metabase-postgres         |  ["--> query_processor.util.add_alias_info$field_source_table_alias.invokeStatic(add_alias_info.clj:182)"
metabase-postgres         |   "query_processor.util.add_alias_info$field_source_table_alias.invoke(add_alias_info.clj:171)"
metabase-postgres         |   "query_processor.util.add_alias_info$fn__44807.invokeStatic(add_alias_info.clj:316)"
metabase-postgres         |   "query_processor.util.add_alias_info$fn__44807.invoke(add_alias_info.clj:313)"
metabase-postgres         |   "query_processor.util.add_alias_info$add_alias_info_STAR_$replace_44843__44844$fn__44851$fn__44852.invoke(add_alias_info.clj:371)"
metabase-postgres         |   "query_processor.util.add_alias_info$add_alias_info_STAR_$replace_44843__44844$fn__44851.invoke(add_alias_info.clj:371)"
metabase-postgres         |   "query_processor.util.add_alias_info$add_alias_info_STAR_$replace_44843__44844.invoke(add_alias_info.clj:371)"
metabase-postgres         |   "mbql.util.match.impl$replace_in_collection.invokeStatic(impl.cljc:47)"
metabase-postgres         |   "mbql.util.match.impl$replace_in_collection.invoke(impl.cljc:38)"
metabase-postgres         |   "query_processor.util.add_alias_info$add_alias_info_STAR_$replace_44843__44844$fn__44851.invoke(add_alias_info.clj:371)"
metabase-postgres         |   "query_processor.util.add_alias_info$add_alias_info_STAR_$replace_44843__44844.invoke(add_alias_info.clj:371)"
metabase-postgres         |   "mbql.util.match.impl$replace_in_collection.invokeStatic(impl.cljc:47)"
metabase-postgres         |   "mbql.util.match.impl$replace_in_collection.invoke(impl.cljc:38)"
metabase-postgres         |   "query_processor.util.add_alias_info$add_alias_info_STAR_$replace_44843__44844$fn__44851.invoke(add_alias_info.clj:371)"
metabase-postgres         |   "query_processor.util.add_alias_info$add_alias_info_STAR_$replace_44843__44844.invoke(add_alias_info.clj:371)"
metabase-postgres         |   "mbql.util.match.impl$replace_in_collection$iter__20188__20192$fn__20193.invoke(impl.cljc:44)"
metabase-postgres         |   "mbql.util.match.impl$replace_in_collection.invokeStatic(impl.cljc:43)"
metabase-postgres         |   "mbql.util.match.impl$replace_in_collection.invoke(impl.cljc:38)"
metabase-postgres         |   "query_processor.util.add_alias_info$add_alias_info_STAR_$replace_44843__44844$fn__44851.invoke(add_alias_info.clj:371)"
metabase-postgres         |   "query_processor.util.add_alias_info$add_alias_info_STAR_$replace_44843__44844.invoke(add_alias_info.clj:371)"
metabase-postgres         |   "query_processor.util.add_alias_info$add_alias_info_STAR_.invokeStatic(add_alias_info.clj:371)"
metabase-postgres         |   "query_processor.util.add_alias_info$add_alias_info_STAR_.invoke(add_alias_info.clj:368)"
metabase-postgres         |   "query_processor.util.add_alias_info$add_alias_info$fn__44861.invoke(add_alias_info.clj:415)"
metabase-postgres         |   "query_processor.util.add_alias_info$add_alias_info.invokeStatic(add_alias_info.clj:410)"
metabase-postgres         |   "query_processor.util.add_alias_info$add_alias_info.invoke(add_alias_info.clj:381)"
metabase-postgres         |   "driver.sql.query_processor$fn__47277.invokeStatic(query_processor.clj:971)"
metabase-postgres         |   "driver.sql.query_processor$fn__47277.invoke(query_processor.clj:969)"
metabase-postgres         |   "driver.sql.query_processor$mbql__GT_honeysql.invokeStatic(query_processor.clj:976)"
metabase-postgres         |   "driver.sql.query_processor$mbql__GT_honeysql.invoke(query_processor.clj:973)"
metabase-postgres         |   "driver.sql.query_processor$mbql__GT_native.invokeStatic(query_processor.clj:987)"
metabase-postgres         |   "driver.sql.query_processor$mbql__GT_native.invoke(query_processor.clj:983)"
metabase-postgres         |   "driver.sql$fn__83942.invokeStatic(sql.clj:40)"
metabase-postgres         |   "driver.sql$fn__83942.invoke(sql.clj:38)"
metabase-postgres         |   "query_processor.middleware.mbql_to_native$query__GT_native_form.invokeStatic(mbql_to_native.clj:13)"
metabase-postgres         |   "query_processor.middleware.mbql_to_native$query__GT_native_form.invoke(mbql_to_native.clj:8)"
metabase-postgres         |   "query_processor.middleware.mbql_to_native$mbql__GT_native$fn__51552.invoke(mbql_to_native.clj:20)"
metabase-postgres         |   "query_processor$fn__54083$combined_post_process__54088$combined_post_process_STAR___54089.invoke(query_processor.clj:212)"
metabase-postgres         |   "query_processor$fn__54083$combined_pre_process__54084$combined_pre_process_STAR___54085.invoke(query_processor.clj:209)"
metabase-postgres         |   "query_processor.middleware.resolve_database_and_driver$resolve_database_and_driver$fn__52528$fn__52533.invoke(resolve_database_and_driver.clj:35)"
metabase-postgres         |   "driver$do_with_driver.invokeStatic(driver.clj:75)"
metabase-postgres         |   "driver$do_with_driver.invoke(driver.clj:71)"
metabase-postgres         |   "query_processor.middleware.resolve_database_and_driver$resolve_database_and_driver$fn__52528.invoke(resolve_database_and_driver.clj:34)"
metabase-postgres         |   "query_processor.middleware.fetch_source_query$resolve_card_id_source_tables$fn__48563.invoke(fetch_source_query.clj:346)"
metabase-postgres         |   "query_processor.middleware.store$initialize_store$fn__48751$fn__48752.invoke(store.clj:11)"
metabase-postgres         |   "query_processor.store$do_with_store.invokeStatic(store.clj:44)"
metabase-postgres         |   "query_processor.store$do_with_store.invoke(store.clj:38)"
metabase-postgres         |   "query_processor.middleware.store$initialize_store$fn__48751.invoke(store.clj:10)"
metabase-postgres         |   "query_processor.middleware.normalize_query$normalize$fn__52800.invoke(normalize_query.clj:22)"
metabase-postgres         |   "metabase_enterprise.audit_app.query_processor.middleware.handle_audit_queries$handle_internal_queries$fn__59625.invoke(handle_audit_queries.clj:130)"
metabase-postgres         |   "query_processor.middleware.constraints$add_default_userland_constraints$fn__49833.invoke(constraints.clj:53)"
metabase-postgres         |   "query_processor.middleware.process_userland_query$process_userland_query$fn__52739.invoke(process_userland_query.clj:145)"
metabase-postgres         |   "query_processor.middleware.catch_exceptions$catch_exceptions$fn__53111.invoke(catch_exceptions.clj:167)"
metabase-postgres         |   "query_processor.reducible$async_qp$qp_STAR___44972$thunk__44974.invoke(reducible.clj:100)"
metabase-postgres         |   "query_processor.reducible$async_qp$qp_STAR___44972.invoke(reducible.clj:106)"
metabase-postgres         |   "query_processor.reducible$async_qp$qp_STAR___44972.invoke(reducible.clj:91)"
metabase-postgres         |   "query_processor.reducible$sync_qp$qp_STAR___44983.doInvoke(reducible.clj:126)"
metabase-postgres         |   "query_processor$process_userland_query.invokeStatic(query_processor.clj:331)"
metabase-postgres         |   "query_processor$process_userland_query.doInvoke(query_processor.clj:327)"
metabase-postgres         |   "query_processor$fn__54130$process_query_and_save_execution_BANG___54139$fn__54142.invoke(query_processor.clj:342)"
metabase-postgres         |   "query_processor$fn__54130$process_query_and_save_execution_BANG___54139.invoke(query_processor.clj:335)"
metabase-postgres         |   "query_processor$fn__54174$process_query_and_save_with_max_results_constraints_BANG___54183$fn__54186.invoke(query_processor.clj:354)"
metabase-postgres         |   "query_processor$fn__54174$process_query_and_save_with_max_results_constraints_BANG___54183.invoke(query_processor.clj:347)"
metabase-postgres         |   "api.dataset$run_query_async$fn__69583.invoke(dataset.clj:69)"
metabase-postgres         |   "query_processor.streaming$streaming_response_STAR_$fn__40090$fn__40091.invoke(streaming.clj:162)"
metabase-postgres         |   "query_processor.streaming$streaming_response_STAR_$fn__40090.invoke(streaming.clj:161)"
metabase-postgres         |   "async.streaming_response$do_f_STAR_.invokeStatic(streaming_response.clj:65)"
metabase-postgres         |   "async.streaming_response$do_f_STAR_.invoke(streaming_response.clj:63)"
metabase-postgres         |   "async.streaming_response$do_f_async$task__19006.invoke(streaming_response.clj:84)"],
metabase-postgres         |  :card_id nil,
metabase-postgres         |  :context :ad-hoc,
metabase-postgres         |  :error "Cannot determine the source table or query for Field clause [:field 30 {:join-alias \"Products\"}]",
metabase-postgres         |  :row_count 0,
metabase-postgres         |  :running_time 0,
metabase-postgres         |  :preprocessed
metabase-postgres         |  {:type :query,
metabase-postgres         |   :database 1,
metabase-postgres         |   :query
metabase-postgres         |   {:source-table 4,
metabase-postgres         |    :breakout
metabase-postgres         |    [[:field 31 {:binning {:strategy :num-bins, :min-value 1.0, :max-value 5.0, :num-bins 8, :bin-width 1/2}}]],
metabase-postgres         |    :aggregation [[:aggregation-options [:count] {:name "count"}]],
metabase-postgres         |    :filter
metabase-postgres         |    [:and
metabase-postgres         |     [:=
metabase-postgres         |      [:field 36 nil]
metabase-postgres         |      [:value
metabase-postgres         |       680N
metabase-postgres         |       {:base_type :type/BigInteger,
metabase-postgres         |        :effective_type :type/BigInteger,
metabase-postgres         |        :coercion_strategy nil,
metabase-postgres         |        :semantic_type :type/PK,
metabase-postgres         |        :database_type "BIGINT",
metabase-postgres         |        :name "ID"}]]
metabase-postgres         |     [:=
metabase-postgres         |      [:field 33 nil]
metabase-postgres         |      [:value
metabase-postgres         |       118
metabase-postgres         |       {:base_type :type/Integer,
metabase-postgres         |        :effective_type :type/Integer,
metabase-postgres         |        :coercion_strategy nil,
metabase-postgres         |        :semantic_type :type/FK,
metabase-postgres         |        :database_type "INTEGER",
metabase-postgres         |        :name "PRODUCT_ID"}]]
metabase-postgres         |     [:=
metabase-postgres         |      [:field 31 nil]
metabase-postgres         |      [:value
metabase-postgres         |       3
metabase-postgres         |       {:base_type :type/Integer,
metabase-postgres         |        :effective_type :type/Integer,
metabase-postgres         |        :coercion_strategy nil,
metabase-postgres         |        :semantic_type :type/Score,
metabase-postgres         |        :database_type "SMALLINT",
metabase-postgres         |        :name "RATING"}]]
metabase-postgres         |     [:=
metabase-postgres         |      [:field 30 {:join-alias "Products"}]
metabase-postgres         |      [:value
metabase-postgres         |       118N
metabase-postgres         |       {:base_type :type/BigInteger,
metabase-postgres         |        :effective_type :type/BigInteger,
metabase-postgres         |        :coercion_strategy nil,
metabase-postgres         |        :semantic_type :type/PK,
metabase-postgres         |        :database_type "BIGINT",
metabase-postgres         |        :name "ID"}]]],
metabase-postgres         |    :order-by
metabase-postgres         |    [[:asc [:field 31 {:binning {:strategy :num-bins, :min-value 1.0, :max-value 5.0, :num-bins 8, :bin-width 1/2}}]]]},
metabase-postgres         |   :middleware {:js-int-to-string? true, :add-default-userland-constraints? true},
metabase-postgres         |   :info {:executed-by 1, :context :ad-hoc}},
metabase-postgres         |  :ex-data
metabase-postgres         |  {:type :invalid-query,
metabase-postgres         |   :clause [:field 30 {:join-alias "Products"}],
metabase-postgres         |   :query
metabase-postgres         |   {:source-table 4,
metabase-postgres         |    :breakout
metabase-postgres         |    [[:field 31 {:binning {:strategy :num-bins, :min-value 1.0, :max-value 5.0, :num-bins 8, :bin-width 1/2}}]],
metabase-postgres         |    :aggregation [[:aggregation-options [:count] {:name "count"}]],
metabase-postgres         |    :filter
metabase-postgres         |    [:and
metabase-postgres         |     [:=
metabase-postgres         |      [:field 36 nil]
metabase-postgres         |      [:value
metabase-postgres         |       680N
metabase-postgres         |       {:base_type :type/BigInteger,
metabase-postgres         |        :effective_type :type/BigInteger,
metabase-postgres         |        :coercion_strategy nil,
metabase-postgres         |        :semantic_type :type/PK,
metabase-postgres         |        :database_type "BIGINT",
metabase-postgres         |        :name "ID"}]]
metabase-postgres         |     [:=
metabase-postgres         |      [:field 33 nil]
metabase-postgres         |      [:value
metabase-postgres         |       118
metabase-postgres         |       {:base_type :type/Integer,
metabase-postgres         |        :effective_type :type/Integer,
metabase-postgres         |        :coercion_strategy nil,
metabase-postgres         |        :semantic_type :type/FK,
metabase-postgres         |        :database_type "INTEGER",
metabase-postgres         |        :name "PRODUCT_ID"}]]
metabase-postgres         |     [:=
metabase-postgres         |      [:field 31 nil]
metabase-postgres         |      [:value
metabase-postgres         |       3
metabase-postgres         |       {:base_type :type/Integer,
metabase-postgres         |        :effective_type :type/Integer,
metabase-postgres         |        :coercion_strategy nil,
metabase-postgres         |        :semantic_type :type/Score,
metabase-postgres         |        :database_type "SMALLINT",
metabase-postgres         |        :name "RATING"}]]
metabase-postgres         |     [:=
metabase-postgres         |      [:field 30 {:join-alias "Products"}]
metabase-postgres         |      [:value
metabase-postgres         |       118N
metabase-postgres         |       {:base_type :type/BigInteger,
metabase-postgres         |        :effective_type :type/BigInteger,
metabase-postgres         |        :coercion_strategy nil,
metabase-postgres         |        :semantic_type :type/PK,
metabase-postgres         |        :database_type "BIGINT",
metabase-postgres         |        :name "ID"}]]],
metabase-postgres         |    :order-by
metabase-postgres         |    [[:asc
metabase-postgres         |      [:field 31 {:binning {:strategy :num-bins, :min-value 1.0, :max-value 5.0, :num-bins 8, :bin-width 1/2}}]]]}},
metabase-postgres         |  :data {:rows [], :cols []}}
```
</details>

**To Reproduce**
1. Go to New Question -> Select Orders Table 
2. Join Product with Orders Table 
3. Create a simple aggregate and make sure to group by a column with the joined table
image

5. Press Visualize -> Chose Line Visualization 
6. Select point on chart and select Compare to the rest

image

5. Observe broken charts and ` Error processing query: Cannot determine the source table or query for Field clause [:field 30 {:join-alias ...` in logs

image

**Expected behavior**

Load data for broken charts. If you press on the card the data loads as expected. Only within the compare dashboard it doesn't load

**Information about your Metabase Installation:**

You can get this information by going to Admin -> Troubleshooting.

- Your browser and the version:  Chrome 52.1
- Your databases:  Sample Database and PostgeSQL
- Metabase version: 0.44.2

### LINKED ISSUE #32557: Compare to the Rest generates dashboard with failing questions
**Describe the bug**
Automatic explorations: Compare to the rest creates a dashboard with several failing questions from the  sample dataset.

**Logs**
[logs.txt](https://github.com/metabase/metabase/files/12129532/logs.txt)

**To Reproduce**
On the sample dataset:
1. Create a question using the query builder: 
-- pick `Orders`, join with `Products` on `Orders.Product ID` = `Products.ID`
-- summarize by count of rows
-- group by `Created At: Month`, `Products.Category`
2. Visualize
3. Click on a data point, and select Compare to the rest
4. A dashboard is created with several failing charts

**Expected behavior**
We should not see any failing charts on the automatically generated dashboard.

**Screenshots**
image

**Severity**
How severe an issue is this bug to you? Is this annoying, blocking some users, blocking an upgrade or blocking your usage of Metabase entirely?
Note: the more honest and specific you are here the more we will take you seriously.

**Additional context**

**Metabase Diagnostic Info**

```json
{
  "browser-info": {
    "language": "en-GB",
    "platform": "MacIntel",
    "userAgent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/114.0.0.0 Safari/537.36",
    "vendor": "Google Inc."
  },
  "system-info": {
    "file.encoding": "UTF-8",
    "java.runtime.name": "OpenJDK Runtime Environment",
    "java.runtime.version": "11.0.19+7",
    "java.vendor": "Eclipse Adoptium",
    "java.vendor.url": "https://adoptium.net/",
    "java.version": "11.0.19",
    "java.vm.name": "OpenJDK 64-Bit Server VM",
    "java.vm.version": "11.0.19+7",
    "os.name": "Linux",
    "os.version": "5.4.209-116.367.amzn2.x86_64",
    "user.language": "en",
    "user.timezone": "GMT"
  },
  "metabase-info": {
    "databases": [
      "h2"
    ],
    "hosting-env": "unknown",
    "application-database": "postgres",
    "application-database-details": {
      "database": {
        "name": "PostgreSQL",
        "version": "14.7"
      },
      "jdbc-driver": {
        "name": "PostgreSQL JDBC Driver",
        "version": "42.5.1"
      }
    },
    "run-mode": "prod",
    "version": {
      "date": "2023-06-14",
      "tag": "v1.46.5",
      "branch": "release-x.46.x",
      "hash": "272e3e6"
    },
    "settings": {
      "report-timezone": null
    }
  }
}
```

### PR DESCRIPTION
**Pull request by markbastian**
In `query-based-analysis` the final stage was to `maybe-enrich-joins` to add in joins defined in the query being x-rayed into the dashcard query. In some cases (e.g. cell queries), the trigger to detect this addition isn't known as the filters are added downstream.

This PR always adds in the join aliases if they are present since they should be harmless if not needed and you can't know if they are needed downstream as the code is currently designed.

The tests assert not only that the cards are created, but that the filter is present and executable (these previously errored out).

Fixes [25278](https://github.com/metabase/metabase/issues/25278)
Fixes [32557](https://github.com/metabase/metabase/issues/32557)

### CODE REVIEW

File: src/metabase/automagic_dashboards/core.clj [hunk 1]
```diff
@@ -779,23 +779,15 @@
...
-  (let [query (get-in card-member [:card :dataset_query :query])]
-    (if (key-in? query :join-alias)
-      ;; Always in the top level even if the join-alias is found deep in there
-      (assoc-in card-member [:card :dataset_query :query :joins] join-statement)
-      card-member)))
-
-(defn- maybe-enrich-joins
+  "If the dashcard is a query card, ensure the joins are present."
+  [join-statement dashcard]
+  (if (get-in dashcard [:card :dataset_query :query])
```
Comment (escherize) [inline src/metabase/automagic_dashboards/core.clj:785]: I think a cond-> could work well here.

File: src/metabase/automagic_dashboards/core.clj [hunk 2]
```diff
@@ -779,23 +779,15 @@
...
-  [join-statement card-member]
-  (let [query (get-in card-member [:card :dataset_query :query])]
-    (if (key-in? query :join-alias)
-      ;; Always in the top level even if the join-alias is found deep in there
-      (assoc-in card-member [:card :dataset_query :query :joins] join-statement)
-      card-member)))
-
-(defn- maybe-enrich-joins
+  "If the dashcard is a query card, ensure the joins are present."
+  [join-statement dashcard]
```
Comment (escherize) [inline src/metabase/automagic_dashboards/core.clj:784]: `splice-in`'s args feel like they are in the wrong order to me. it is kind of an assoc, so I would have expected the thing you are assoc-ing onto to come first.

**Review by escherize — APPROVED** [review]
The tests test that there are actually join aliases always there. lgtm

File: src/metabase/automagic_dashboards/core.clj (hunk 1 above)
Comment (markbastian) [reply src/metabase/automagic_dashboards/core.clj:785]: Done.

File: src/metabase/automagic_dashboards/core.clj (hunk 2 above)
Comment (markbastian) [reply src/metabase/automagic_dashboards/core.clj:784]: Once you collapse the splice fn down to one form it makes sense to just collapse it all down into `preserve-joins`.

**Pull request merged**